internal/service/ecs/cluster_capacity_providers.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/hashicorp/terraform-provider-aws/internal/conns"
+	"github.com/hashicorp/terraform-provider-aws/internal/flex"
+	"github.com/hashicorp/terraform-provider-aws/internal/tfresource"
+)
+
+func ResourceClusterCapacityProviders() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceClusterCapacityProvidersPut,
+		ReadContext:   resourceClusterCapacityProvidersRead,
+		UpdateContext: resourceClusterCapacityProvidersPut,
+		DeleteContext: resourceClusterCapacityProvidersDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"capacity_providers": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"cluster_name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				// The API accepts both an ARN and a name in a generic "cluster"
+				// parameter, but allowing that would force the resource to guess
+				// which one to return on read.
+				ValidateFunc: validateClusterName,
+			},
+			"default_capacity_provider_strategy": {
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"base": {
+							Type:         schema.TypeInt,
+							Default:      0,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(0, 100000),
+						},
+						"capacity_provider": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"weight": {
+							Type:         schema.TypeInt,
+							Default:      0,
+							Optional:     true,
+							ValidateFunc: validation.IntBetween(0, 1000),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceClusterCapacityProvidersPut(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).ECSConn
+
+	clusterName := d.Get("cluster_name").(string)
+
+	input := &ecs.PutClusterCapacityProvidersInput{
+		Cluster:                         aws.String(clusterName),
+		CapacityProviders:               flex.ExpandStringSet(d.Get("capacity_providers").(*schema.Set)),
+		DefaultCapacityProviderStrategy: expandCapacityProviderStrategy(d.Get("default_capacity_provider_strategy").(*schema.Set)),
+	}
+
+	log.Printf("[DEBUG] Updating ECS cluster capacity providers: %s", input)
+
+	err := retryClusterCapacityProvidersPut(ctx, conn, input)
+
+	if err != nil {
+		return diag.Errorf("error updating ECS Cluster (%s) capacity providers: %s", clusterName, err)
+	}
+
+	if _, err := waitClusterAvailable(ctx, conn, clusterName); err != nil {
+		return diag.Errorf("error waiting for ECS Cluster (%s) to become available: %s", clusterName, err)
+	}
+
+	d.SetId(clusterName)
```

Comment:
Do you foresee problems with only using the cluster name as the ID? Could someone validly use multiple of these resources with the same cluster?

Reply:
As updating the capacity providers is always all or nothing, doing so from multiple locations would lead to perpetual configuration drift. I'll add a documentation note to that effect.